Add function to obtain the waveform on a PulseChannel

snsunx opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

I have written a function `get_channel_waveform` in `pulse/transforms/channel_transforms.py` to obtain the waveform on a PulseChannel.


### Details and comments

I have tested the function on predefined pulse schedules, specifically the `cx01` and `meas0` in the following code. 

```python
IBMQ.load_account()
provider = IBMQ.get_provider('redacted')
backend = provider.get_backend('ibmq_bogota')
inst_sched_map = backend.defaults().instruction_schedule_map
cx01 = inst_sched_map.get('cx', (0, 1))
meas0 = inst_sched_map.get('measure', 0)
```

It feels to me that the function should be a method of `Schedule` rather than a standalone function in a standalone script. I would like to know what the maintainers think.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=6995) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=6995) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=6995) it.</sub>

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+from qiskit.pulse.channels import PulseChannel
+from qiskit.visualization.pulse_v2.events import ChannelEvents
+
+def get_channel_waveform(sched: Schedule, 
```
Should this also accept a `ScheduleBlock` @nkanazawa1989?

nkanazawa1989 replied:
Not needed if it is correctly implemented. This transform pass internally calls full `target_qobj_transform`, however, we need a specific transform function that takes a low-level representation (all directives are dropped, contexts are resolved, subroutine is flattened, etc...) and generate a raw waveform data by parsing frames. In this case, the the input program should be `Schedule`.

snsunx replied:
The `target_qobj_transform` function has the input annotation as `sched: Union[ScheduleBlock, Schedule, InstructionSched, Iterable[InstructionSched]]`. Maybe I should just do the same here?

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+    """
+    # Check consistency of arguments
+    if not isinstance(chan, PulseChannel):
+        raise TypeError("The channel must be a DriveChannel, " 
+                        "ControlChannel or a MeasureChannel")
```
```suggestion
        raise TypeError("The channel must be a PulseChannel eg., a DriveChannel, " 
                        "ControlChannel or MeasureChannel")

```

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+        raise TypeError("The channel must be a DriveChannel, " 
+                        "ControlChannel or a MeasureChannel")
+    if apply_carrier_wave:
+        if backend is not None and qubit_index is not None:
```
Priority should be to use `chan_freq` if supplied as override and fall back to the `backend` if `chan_freq` is not supplied.

snsunx replied:
Ok I will address this in the argument checking section

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+                        "ControlChannel or a MeasureChannel")
+    if apply_carrier_wave:
+        if backend is not None and qubit_index is not None:
+            if isinstance(chan, MeasureChannel):
+                chan_freq = backend.defaults().meas_freq_est[qubit_index]
+            else:
+                chan_freq = backend.defaults().qubit_freq_est[qubit_index]
```
I think we should separate this out as a method. We will also need to look up the frequency for the `ControlChannel` based on the `u_channel_lo`. I think this must be happening somewhere already. Any ideas @nkanazawa1989?

taalexander replied:
if `apply_carrier_wave` is supplied and we cannot calculate the `chan_freq` from the backend, eg., not a drive, control or measure channel we should raise an error and force the user to supply `chan_freq` via an informative message.

nkanazawa1989 replied:
Yes, there is a backend class parser here. We need similar thing for this.
https://github.com/Qiskit/qiskit-terra/blob/117d55df8933ca79753d01f14483267bc7e6112a/qiskit/visualization/pulse_v2/device_info.py#L95

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+from qiskit.providers.ibmq import IBMQBackend
+from qiskit.pulse import Schedule, Play, MeasureChannel
+from qiskit.pulse.channels import PulseChannel
+from qiskit.visualization.pulse_v2.events import ChannelEvents
```
We likely should not be using ChannelEvents in the transforms module (circular imports). This may need to be moved, or is there another work around @nkanazawa1989?

nkanazawa1989 replied:
That was my intention. If we separate `ChannelEvents` from the visualizer, we can use that functionality for other purpose as well. Just porting the class from visualizer is not right way though it works. This approach may conflict with #5977

nkanazawa1989 replied:
I'm fine with deprecating `ChannelEvents` and writing new one in the transforms. This will drastically simplify the visualizer.

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+    sched_trans = target_qobj_transform(sched)
+    chan_events = ChannelEvents.load_program(sched_trans, chan)
+    waveform_inst_tups = chan_events.get_waveforms()
+    if backend is not None:
```
Fallback to backend if dt is not supplied.

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+            
+            # Apply phase and frequency shifts and optionally carrier wave
+            pulse_waveform = inst_tup.inst.pulse.get_waveform().samples
+            pulse_waveform *= np.exp(1j * phase)            
+            pulse_waveform *= np.exp(1j * freq * t_array)
```
```suggestion
            pulse_waveform *= np.exp(1j * (2 * np.pi* freq * t_array + phase))

```

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+            pulse_waveform *= np.exp(1j * phase)            
+            pulse_waveform *= np.exp(1j * freq * t_array)
+            if apply_carrier_wave:
+                pulse_waveform *= np.exp(1j * chan_freq * t_array)
```
```suggestion
                pulse_waveform *= np.exp(1j * 2 * np.pi * chan_freq * t_array)

```

taalexander commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+            if apply_carrier_wave:
+                pulse_waveform *= np.exp(1j * chan_freq * t_array)
+            chan_waveform[t0:tf] += pulse_waveform
+    return chan_waveform
```
```suggestion
    return chan_waveform

```

taalexander reviewed (changes requested):
This is a great start. Could you also please add tests for this functionality? I would also appreciate if @nkanazawa1989 could take a look at this.

nkanazawa1989 reviewed (changes requested):
Thanks for taking a stab at waveform transform. I'm grad to see we have another transform function. My two major points are

- Move frame parser to transform
  - Transform should not have dependency on visualizer. You can create new parser in the pulse transform module, then use new transform function in the visualizer.

- Split function into multiple pieces
  - One huge transform function is usually hard to maintain. If we split the transform into multiple [atomic functions](https://www.codementor.io/@seantullis/improve-your-code-with-atomic-functions-r6dt43fy7), we can efficiently write unittest. For example,
    - Just call get waveform if any parametric pulse exist (`to_waveform`)
    - Apply phase shift to each waveform (`apply_phase`)
    - Apply carrier waves based on frequency at t0 (`apply_carrier`)
    - Combine all waveforms in the channel to create full waveform array (`to_full_waveform`)
  - By combining these passes with `target_qobj_transform`, you can create new transform function, something like `waveform_transform`. For the visualizer, you can create another transform function that only performs `to_waveform` and `apply_phase`. Note that the visualizer draws pulses individually. 

> It feels to me that the function should be a method of Schedule rather than a standalone function in a standalone script. I would like to know what the maintainers think.

This is usability vs maintenance issue. I think `Schedule` is a container of instructions, i.e. a program, and thus the operations (i.e. transforms) and programs should be separated. In Qiskit, `QuantumCircuit` doesn't take this approach and it has many method to transform itself, however, now the class has nearly 5000 lines of code. The number of lines will be further increased if we add new transform. 

This is not preferable in terms of test and code maintenance. If we separate operations from the program, we can implement as many transforms as we want while keeping the size of `Schedule` class. Any thoughts? @taalexander

snsunx commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+def get_channel_waveform(sched: Schedule, 
+                         chan: PulseChannel, 
+                         backend: Union[None, IBMQBackend] = None,
+                         qubit_index: Union[None, int] = None,
```
Do you mean I should put the type annotations as `Optional[IBMQBackend]` etc?

snsunx replied:
Do you mean `qubit_index` is not required? Now I can see it can be obtained from `chan.index`.

snsunx commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,86 @@
+
+def get_channel_waveform(sched: Schedule, 
+                         chan: PulseChannel, 
+                         backend: Union[None, IBMQBackend] = None,
```
Ok I will include some error flags in the argument checking section

snsunx commented:
Thank you two for your comments. I have updated some argument type checking statements, which I think addressed most of @taalexander's comments. One remaining task is to figure out the correct frequency for a `ControlChannel`. Additionally, since I'm new to Qiskit development, can I ask where I should include the unit tests? Is there an example script I can look at?

I will take a look at how to port the functionalities of `ChannelEvents` over to the `pulse/transform` module, which may take a bit of time to figure out.

nkanazawa1989 commented:
[Here](https://github.com/Qiskit/qiskit-terra/blob/main/test/python/pulse/test_transforms.py) you can find some unittests for the pulse transforms. Please let me know if you have any question about the implementation of the `ChannelEvents`. The frequency of control channel is computed in the https://github.com/Qiskit/qiskit-terra/blob/117d55df8933ca79753d01f14483267bc7e6112a/qiskit/visualization/pulse_v2/device_info.py#L95

Note that according to the [OpenPulse specification](https://arxiv.org/abs/1809.03452), control channel can be whatever we can describe as a device Hamiltonian. This means, the frequency computation of this channel may depend on the device architecture. So you need to write a base class that implements abstract method to compute control channel frequencies, along with its subclass that implements a specific computation for IBM devices.

snsunx commented:
@nkanazawa1989 Thank you! Sorry I have been busy with some other work and haven't been able to get to this. I should be able to give another update later this week.

snsunx commented:
Hi, sorry for a slightly late update. I moved `ChannelEvents` to `pulse.transforms` and renamed it to `ChannelTransforms`. The new function `get_parsed_instructions` is mostly based on `get_waveforms` in the `ChannelEvents` class but also includes three additional steps: `_parse_waveform`, `_apply_phase`, and optionally `_apply_frequency`. I still need to work on integrating it with the plotting functions under the `visualization` module along with some cleanup. I will also write some unit tests later. Meanwhile if any of you has suggestions please let me know.

nkanazawa1989 commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,330 @@
+ScheduleLike = Union[pulse.Schedule, pulse.ScheduleBlock, 
+                     InstructionSched, Iterable[InstructionSched]]
+
+class ParsedInstruction:
```
You can use [dataclass](https://docs.python.org/3/library/dataclasses.html) just to store a set of data.

nkanazawa1989 commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,330 @@
+from device_info import OpenPulseBackendInfo
+
+InstructionSched = Union[Tuple[int, Instruction], Instruction]
+ScheduleLike = Union[pulse.Schedule, pulse.ScheduleBlock, 
```
I like this :)

nkanazawa1989 commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,330 @@
+class ChannelTransforms:
+    """Channel transform manager."""
+
+    _waveform_group = (pulse.Play, pulse.Delay, pulse.Acquire)
```
Maybe we can move this to below the `ScheduleLike` so that it can be easily found (when we add new instruction)?

nkanazawa1989 commented on qiskit/pulse/transforms/channel_transforms.py:
```diff
@@ -0,0 +1,330 @@
+            The channel transform manager for the specified channel.
+        """
+        # flatten the schedule
+        program = target_qobj_transform(program)
```
We can assume this is externally done by writing another transform function including this class like a pass. Otherwise this transform will have unnecessary dependency on `target_qobj_transform`.

nkanazawa1989 commented on qiskit/pulse/transforms/device_info.py:
```diff
@@ -0,0 +1,145 @@
+from qiskit.providers import BaseBackend, BackendConfigurationError
+
+
+class DrawerBackendInfo(ABC):
```
Should be renamed since this is no longer drawer.

nkanazawa1989 commented on qiskit/pulse/transforms/pulse_instruction_types.py:
```diff
@@ -0,0 +1,236 @@
+from qiskit import pulse
+
+
+class PhaseFreqTuple(NamedTuple):
```
You can update these named tuples with dataclass since it's more readable (when I wrote the original code I was not able to use dataclass since we needed to support python3.6).

nkanazawa1989 reviewed (commented):
Thanks @snsunx for the constant update! I think now it is moving toward right direction. You can also create a dedicated folder in the pulse transforms since this transform creates several files. And you need to move some data structure back to drawer, e.g. `class LineType(str, Enum):`. 

I'll convert this PR :)

snsunx commented:
Hi @nkanazawa1989, I didn't create a new folder but made the following organizational changes:

- I put `PhaseFreqTuple` and `ParsedInstruction` in `channel_transforms.py`. I was thinking maybe putting them in a separate script, but they seem to be only used in `ChannelTransforms` and I think fewer scripts are easier to maintain
- I moved the script `backend_info.py` which contains `BackendInfo` and `OpenPulseBackendInfo` to `qiskit/providers/models`
 
I wonder what are your thoughts on these changes. If these changes are ok, I'll go on to write some unit tests and also modify the existing unit tests in `test/python/visualization/test_generators.py`, which is the reason for some of the failed checks.

Also I made a forced amend push (from `30125c1` to `7c9fe84`), but I realized it's a bad idea. I'll avoid doing that in the future.

nkanazawa1989 commented:
> I put PhaseFreqTuple and ParsedInstruction in channel_transforms.py. I was thinking maybe putting them in a separate script, but they seem to be only used in ChannelTransforms and I think fewer scripts are easier to maintain

Yes, that should be okey. 

> I moved the script backend_info.py which contains BackendInfo and OpenPulseBackendInfo to qiskit/providers/models. I removed "Drawer" from the name DrawerBackendInfo since the class itself doesn't seem to have direct implications to drawer functions

I don't think this is good direction. The `qiskit.providers` is a module that defines a template of backend and provider model so that third party (i.e. non-IBM) provider can use these classes as a baseclass of their cloud and backend API. `BackendInfo` doesn't provide such API and this is just a parser of it. If this class is used only by the transform the file should be in the transform module, but if you find this class is also useful for other pulse functionality probably you can put the file in `qiskit.pulse`.

nkanazawa1989 commented:
> I made minimal changes to the functions from the original visualization module to avoid accidentally breaking something in the code. I indeed think there are a few more places that can be improved in the module, and it may be better to open another PR for that.

Yes, agreed. The `qiskit.pulse.transforms` is replica of [transpiler passes](https://github.com/Qiskit/qiskit-terra/tree/main/qiskit/transpiler/passes), and I intend to keep capability of integration of this chain into the pass manager. However, the `ChannelTransforms` class is no longer a pass because of two reasons.
- We cannot divide functionality. Always need to add entire functionality with disabling unused part. This doesn't fit in with the concept of passes.
- I don't want to introduce another representation of schedule, i.e. `ParsedInstruction`. This is intended to be in visualization. The data representation existing in `qiskit.pulse` should be operable as a schedule. However, this is just a list of dataclass and not compatible with `Schedule` class. We recently added `ScheduleBlock` to the pulse module, but this already causes some complexity in the stack.

Back to your original motivation, probably you just need to add `get_waveform` method to original `ChannelEvents`. Since this method already exist for different purpose, you can add `to_single_waveform` helper function that takes `ChannelEvents` instance as an input. You can create `util.py` file under the `qiskit.visualization.pulse_v2` for this.

Creation of these passes will be done in follow-up PRs if you will :)

snsunx commented:
Now I see that this class should not be in `qiskit.pulse.transforms`. I will open a new PR for this issue starting from the current `main` branch.

> Back to your original motivation, probably you just need to add get_waveform method to original ChannelEvents. Since this method already exist for different purpose, you can add to_single_waveform helper function that takes ChannelEvents instance as an input. You can create util.py file under the qiskit.visualization.pulse_v2 for this.

Should I add the `to_single_waveform` in a new script `utils.py` under `qiskit.visualization.pulse_v2`, or should I add it as a method in `ChannelEvents` as done in this PR? Currently there is no `utils.py` under `qiskit.visualization.pulse_v2` so I'm not sure if it's a good idea to have a separate script for this function.

nkanazawa1989 commented:
Of course you can add the method to the `ChannelEvents` under different name.
